### DISCUSSION TITLE
Inject a service class into a projection

### DISCUSSION THREAD

**Discussion by ronikurnia1** [JasperFx/marten#2329] [Q&A]
I think this is powerful if we can inject a service class (using DI) into a projection, something like this:

```
public class MyProjection : SingleStreamAggregation<MyReadModel>
{
    private readonly IAggregateRepository repository;
    public MyProjection (IAggregateRepository repository)
    {
        this.repository = repository;
    }
}
```

**Comment by jeremydmiller** [comment]
@ronikurnia1 You can use DI services in projections as is. Check the flavor of AddMarten that takes in a `Func<IServiceProvider, StoreOptions>` argument to add projections with service dependencies. You'd be completely responsible for any service lifecycle

**Reply by ronikurnia1** [reply]
@jeremydmiller thanks for the response.
I've tried the following code:

```
builder.Services.AddMarten(options =>
{
    options.Projections.Add<MyProjection>();
}
```
But Matren rejected it, asking for a public parameter-less constructor class

Did I miss something here?

**Reply by gfoidl** [reply]
Add an instance of the projection instead of defining it with the generic type.

E.g.
```c#
appBuilder.Services.AddMarten(serviceProvider =>
{
    StoreOptions options = new();

    // Configuration as usual

    MyProjection myProjection = new(serviceProvider.GetRequiredService<IMyServiceToUseInProjection>());
    options.Projections.Add(myProjection, ProjectionLifecycle.Async);

    return options;
});


public class MyProjection : SingleStreamProjection<MyView>
{
    public MyProjection(IMyServiceToUseInProjection foo)
    {
        // ...
    }
}
```
